### PR TITLE
Add support for GaomonM7 and make improvement for GaomonM6、1060pro、M8

### LINKED ISSUE #3341: Add support for Gaomon M7
### Manufacturer

Gaomon

### Product Name

M7

### Device Width

258.4

### Device Height

161.5

### Pressure Levels

8191 in Gaomon official driver

### Lines per Inch (LPI)

5080

### Additional Information

It is said to have 16384 pressure levels when you open official site.  
I just found that i can only  get 8191 in its official driver like this :
image

**2024.6.19 added**  
This tablet has 13 buttons plus16 virtual buttons: 
- Red: Physical buttons
- Green: Vitual buttons
image

### Diagnostic Information

[2024.6.17export.txt](https://github.com/user-attachments/files/15866320/2024.6.17export.txt)

### Acknowledgements

- [X] I have searched the existing issues and this new issue is not a duplicate of any.
- [X] I have written a concise and meaningful title.

### PR DESCRIPTION
**Pull request by RainyLave**
1. Add supports for GaomonM7 
2. Make improvement for GaomonM6、1060pro、M8：
Models of the M6 include: OEM02_T183 and GM001_T223;
Models of the M7 include: GM001_T207、GM001_T202、GM001_T220;
Models of the M8 include: GM001_T208、GM001_T201、GM001_T221;
Models of the 1060pro include: OEM02_T174、GM001_T213、GM001_T216;
refered by Gaomon Official information
[2024.6.17export.txt](https://github.com/user-attachments/files/17004662/2024.6.17export.txt)
[DevImg.zip](https://github.com/user-attachments/files/17004667/DevImg.zip)
Closes #3341

### CODE REVIEW

**Comment by RainyLave** [comment]
The log while opentabletdrive runs is as below:
[RootHub:Debug]	Initialized internal child hubs: WinUSBRootHub, HidSharpDeviceRootHub
[Detect:Warning]	'Gaomon' driver is detected. It will block detection of tablets. If any problems arise, visit 'https://opentabletdriver.net/Wiki/FAQ/Windows'.
[Detect:Info]	Searching for tablets...
[Detect:Debug]	Attempting to match config 'Artisul A1201' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Gaomon 1060 Pro' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Gaomon M6' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Gaomon M7' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Gaomon M8' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H1060P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H1161' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H420X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H430P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H580X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H610 Pro V2' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H610X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H640P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion H950P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion HC16' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion HS610' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion RTE-100' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Huion RTM-500' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col01#7&2eb30748&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd)'
[Detect:Debug]	Attempting to match config 'Artisul A1201' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_00#7&18fb743c&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon 1060 Pro' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_00#7&18fb743c&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon M6' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_00#7&18fb743c&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon M7' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_00#7&18fb743c&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Found 'Gaomon M7' digitizer: 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_00#7&18fb743c&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Artisul A1201' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon 1060 Pro' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon M6' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon M7' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon M8' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H1060P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H1161' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H420X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H430P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H580X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H610 Pro V2' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H610X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H640P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H950P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion HC16' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion HS610' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion RTE-100' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion RTM-500' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_02&col02#7&2eb30748&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Artisul A1201' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon 1060 Pro' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon M6' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon M7' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Gaomon M8' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H1060P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H1161' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H420X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H430P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H580X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H610 Pro V2' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H610X' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H640P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion H950P' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion HC16' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion HS610' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion RTE-100' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Debug]	Attempting to match config 'Huion RTM-500' to device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_01#7&adbc986&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Detect:Info]	Found tablet 'Gaomon M7'
[Device:Debug]	Initializing device 'Gaomon Tablet_M7 (\\?\hid#vid_256c&pid_0064&mi_00#7&18fb743c&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})'
[Device:Debug]	Using report parser type 'OpenTabletDriver.Configurations.Parsers.UCLogic.UCLogicV2ReportParser'
[Device:Debug]	Initialized string index 200
[Display:Debug]	DPI Awareness enabled
[Display:Debug]	DPI Awareness enabled
[Display:Debug]	DPI Awareness enabled
[AbsoluteMode:Debug]	Input: [258.465x161.535@<129.2325, 80.7675>:0°]
[AbsoluteMode:Debug]	Output: [1920x1080@<960, 540>]
[AbsoluteMode:Debug]	LockAspectRatio: False
[AbsoluteMode:Debug]	AreaClipping: True
[AbsoluteMode:Debug]	AreaLimiting: True
[AbsoluteMode:Debug]	LockToBounds: True
[Gaomon M7:Info]	Output mode: Absolute Mode
[MouseBinding:Debug]	Button: Left
[Settings:Info]	Driver is enabled.
[IPC:Debug]	Connected to a client.
[Display:Debug]	DPI Awareness enabled
So it seems run well with Gaomon M7

**Comment by RainyLave** [comment]
image

**Review by jamesbt365 — CHANGES_REQUESTED** [review]
I do not like the idea of blindly adding support for tablets without properly testing them, just because the official documentation mentions that these strings are used on these tablets does not mean that they are used with these VID and PIDs.

Alongside this, the regex strings can also be improved.

You are welcome to add support for the Gaomon M7 provided you own that tablet and have verified all its specs, but blindly adding strings and changing the max pressure of the Gaomon M8 without testing it on both variants of the tablet is unacceptable IMO.

You may still correct its name that Infinity missed when approving my PR.

In short, do not make changes that you cannot verify functionality for.
